src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -273,9 +289,9 @@ bool QueueIsEmpty ()
 			return _methods.Count == 0;
 		}
 
-		protected virtual void EnqueueMethod (MethodDefinition method)
+		protected virtual void EnqueueMethod (MethodDefinition method, DependencyInfo reason)
 		{
-			_methods.Enqueue (method);
+			_methods.Enqueue ((method, reason));
```

Comment:
This is not good. Please make this code memory neutral to what is today when tracing is disabled

Reply:
This should not allocate new objects - I thought that `(var, var)` is implemented as value types and so this will make the queue slightly larger, but it should not cause new objects on its own. Or are you worried about the size of the queue itself? (Maybe we should measure the max size of the queue when processing some largish app to get a feel of how bad it would be).

Reply:
I agree this needs measurement but there can be a lot of items in the queue (thousands) and this change increases the underlying array size for simple entry from 8 bytes about 3 times if tracing is disabled (default mode)

Reply:
1000s would not worry me too much (we're talking KBs of memory - although if we cross the large-object limit that would not be ideal). That said it might be possible to do without passing the reason into the process method. It's really only used for better tracking and it might be possible to do that in here before we enqueue it. We might still need some kind of a hashtable of methods which were marked like this to get the desired behavior, but that should be a smaller structure. That said I don't think it's possible to get the exact same behavior without at least some additional memory.

Reply:
Let's measure it it's possible that with netcore we get into a lot higher numbers (+ all the reallocations)

Reply:
I've measured a max queue length of 29 for a small console app, 8361 for a new blazor app, and 174397 (!) for a new aspnet app. This seems to be because aspnet is rooting everything, and we put everything rooted onto the queue in one pass before processing any of it. That seems like something we might want to address separately if we're concerned about the memory consumption of the queue.

Reply:
Can you please file a bug for the asp.net case? That number will cause all kinds of issues as it's way too big. I know that in theory the queue can be of any length, but we should try to fix the cases we impose on ourselves :-)

Reply:
Filed https://github.com/mono/linker/issues/1018.